lib/fast_jsonapi/helpers.rb

```diff
@@ -6,7 +6,15 @@ class << self
     # @param [Array<Object>] *params any number of parameters to be passed to the Proc
     # @return [Object] the result of the Proc call with the supplied parameters
     def call_proc(proc, *params)
-      proc.call(*params.take(proc.parameters.length))
+      # The parameters array for a lambda created from a symbol (&:foo) differs
+      # from explictly defined procs/lambdas, so we can't deduce the number of
+      # parameters from the array length. Instead, look for an unnamed
+      # parameter in the first position just pass the record as the receiver obj.
+      if proc.parameters.positive? && !proc.parameters.empty?
+        proc.call(params.first)
+      else
+        proc.call(*params.take(proc.parameters.length))
+      end
```

Comment:
Hope you don't mind the rando leaving a comment here but I don't think `.empty?` is a correct check. On Ruby 2.7, this is what I see:

```
irb(main):056:0> :foo.to_proc.parameters.empty?
=> false
irb(main):057:0> -> {}.parameters.empty?
=> true
irb(main):058:0> ->(a) {}.parameters.empty?
=> false
```

There's also not a `positive?` method on `parameters`. I think it'd be better to say `if proc.arity.negative?` since it appears that the arity of a pretzel-operator/`to_proc` proc is -1

Reply:
Here's the doc for `arity`: https://ruby-doc.org/core-3.0.0/Proc.html#method-i-arity

Reply:
Even better, I just looked at the CRuby source.

For symbol block types, [the min arity is 1 and the max arity is UNLIMITED_ARGUMENTS which is -1](https://github.com/ruby/ruby/blob/8da7f4abc779c6549833d718336d75aac98494b8/proc.c#L1081). The min/max arity is then taken by [rb_proc_arity](https://github.com/ruby/ruby/blob/8da7f4abc779c6549833d718336d75aac98494b8/proc.c#L1123) and this complicatedly written logic runs:

```c
return (proc->is_lambda ? min == max : max != UNLIMITED_ARGUMENTS) ? min : -min-1;
```

symbol blocks are lambdas, but min != max, so this becomes:


```c
return(false ? min : -min-1)
```

So I'd expect on Ruby 3.0 that your arity would be -2 and my arity is -1, so `.arity.negative?` is compatible for both.

Reply:
Excellent, thanks! I couldn't quite fathom how the `.arity` logic worked so I didn't quite trust just checking for negative arity there. I'll simplify the pull request.

Reply:
Me neither, and honestly I'm not sure why they wrote it this way. `-1` would mean "no required arguments and unlimited rest arguments", where `-2` means "one required argument and unlimited rest arguments", which is really counterintuitive. From that perspective, `-2` is actually the correct value because symbol block types require one argument (the receiver), and `-1` is wrong, but it probably shouldn't require reviewing the C source to figure that out.

The right answer would be to make a proc method like `symbol?` for this special case (because this Proc is a special case everywhere anyways in C land - it argument errors by saying "no receiver given" rather than "wrong number of arguments", for instance).

As it stands, the only way I can see to tell this specific situation is to parse `to_s`:

```c
      case block_type_symbol:
	rb_str_catf(str, "%p(&%+"PRIsVALUE")", (void *)self, block->as.symbol);
	break;
```

```irb
irb(main):075:0> :testing.to_proc.to_s
=> "#<Proc:0x00005568af9c9600(&:testing)>"
```

Reply:
This particular case here is pretty specific: on the invoker side there's only a max of two parameters available to send: 

1. The object being serialized
2. The `params` hash

So, there's really no reason for anyone to muck around with splatted parameters in their proc/lambda. I'm not too fussed about having it detect any negative arity (i.e., unlimited parameters) and just pass the object being serialized. That works for the pretzel-operator lambda for both 2.x and 3, and doesn't really detract from any other sensible case.